# Проект : SQL.

# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

# Описание данных

Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задачи исследования


# Изучение общей информации
- загрузить данные из датасетов
- изучить общую информацию о полученных данных

# Выполнить задания
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Вывод

## Изучение общей информации

Импортирую библиотеки. Считываю данные из csv-файла и сохраняю в переменную.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
           SELECT *
           FROM books
        '''

books = pd.io.sql.read_sql(query, con = engine)
books.to_csv('books.csv', index=False)
books.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
           SELECT *
           FROM authors
        '''

authors = pd.io.sql.read_sql(query, con = engine)
authors.to_csv('authors.csv', index=False)
authors.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [5]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
           SELECT *
           FROM publishers
        '''

publishers = pd.io.sql.read_sql(query, con = engine)
publishers.to_csv('publishers.csv', index=False)
publishers.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [6]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
           SELECT *
           FROM ratings
        '''

ratings = pd.io.sql.read_sql(query, con = engine)
ratings.to_csv('ratings.csv', index=False)
ratings.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [7]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
           SELECT *
           FROM reviews
        '''

reviews = pd.io.sql.read_sql(query, con = engine)
reviews.to_csv('reviews.csv', index=False)
reviews.head(10)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Выполнить задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''
           SELECT COUNT(book_id)
           FROM books
           WHERE publication_date >= '2000-01-01'
        '''

count_books = pd.io.sql.read_sql(query, con = engine)
count_books

count
0    821

После 1 января 2000 года вышло 821 книга

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [9]:
query = '''
           SELECT bk.book_id,
                  bk.title,
                  COUNT(DISTINCT review_id) as cnt,
                  AVG(rating) as mean
           
           FROM books as bk
           
           LEFT JOIN reviews AS rev ON bk.book_id = rev.book_id
           LEFT JOIN ratings AS rat ON bk.book_id = rat.book_id
           
           GROUP BY bk.book_id 
           ORDER BY 3 DESC, 4 DESC
       ''' 

rev_rat = pd.io.sql.read_sql(query, con = engine)
rev_rat

book_id                                              title  cnt      mean
0        948                            Twilight (Twilight  #1)    7  3.662500
1        302  Harry Potter and the Prisoner of Azkaban (Harr...    6  4.414634
2        299  Harry Potter and the Chamber of Secrets (Harry...    6  4.287500
3        656                                     The Book Thief    6  4.264151
4        734                                   The Glass Castle    6  4.206897
..       ...                                                ...  ...       ...
995      191  Disney's Beauty and the Beast (A Little Golden...    0  4.000000
996      387                               Leonardo's Notebooks    0  4.000000
997      221                          Essential Tales and Poems    0  4.000000
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel    0  3.666667
999      808                            The Natural Way to Draw    0  3.000000

[1000 rows x 4 columns]

Больше всего обзоров на Twilight (Twilight #1)

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query = '''
           WITH
           
           tab as
           
           (SELECT *
           FROM books as bk
           LEFT JOIN publishers AS pub ON bk.publisher_id  = pub.publisher_id
           WHERE num_pages >= 50)
           
           SELECT publisher,
                  COUNT(book_id) as cnt
           FROM tab
           GROUP BY publisher
           ORDER BY cnt DESC LIMIT 1
        '''

pub = pd.io.sql.read_sql(query, con = engine)
pub

publisher  cnt
0  Penguin Books   42

Издательство, которое выпустило наибольшее число книг толще 50 страниц это Penguin Books

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''
           WITH
           
           tab as
           
           (SELECT title,
                  COUNT(rating) as cnt_r,
                  AVG(rating) as avg_r
           FROM books as bu
           LEFT JOIN ratings as rat ON bu.book_id = rat.book_id
           GROUP BY title)
           
           SELECT author,
                  AVG(avg_r) mean
           FROM tab as ta
           LEFT JOIN books AS bu ON ta.title = bu.title
           LEFT JOIN author AS au ON bu.author_id = au.author_id
           
           WHERE cnt_r >= 50
           GROUP BY author
           ORDER BY mean DESC LIMIT 1
           
        '''
aut = pd.io.sql.read_sql(query, con = engine)
aut

author      mean
0  J.K. Rowling/Mary GrandPré  4.283844

Автор с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками) это J.K. Rowling/Mary GrandPré

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
            WITH
                       
            tab as
            
            (SELECT rat.username as users,
                   COUNT(rating_id) as cnt_rating
            FROM books AS bk
            LEFT JOIN ratings AS rat ON bk.book_id  = rat.book_id
            GROUP BY rat.username),
            
            tab1 as
            
            (SELECT users
            FROM tab 
            WHERE cnt_rating > 50),
            
            tab2 as
            
            (SELECT rev.username as name,
                   COUNT(review_id) as cnt
            FROM tab1 as ta
            LEFT JOIN reviews AS rev ON ta.users  = rev.username
            GROUP BY rev.username)
            
            SELECT ROUND(AVG(cnt))
            FROM tab2
        '''

rev_cnt = pd.io.sql.read_sql(query, con = engine)
rev_cnt

round
0   24.0

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24

## Вывод

По итогам выполнения заданий, были получены следующие результаты:

- После 1 января 2000 года вышло 821 книга
- Больше всего обзоров на Twilight (Twilight #1)
- Издательство, которое выпустило наибольшее число книг толще 50 страниц это Penguin Books
- Автор с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками) это J.K. Rowling/Mary GrandPré
- Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24